openssl/evp.go

```diff
@@ -393,3 +393,34 @@ func newEvpFromParams(id C.int, selection C.int, params []C.OSSL_PARAM) (C.GO_EV
 	}
 	return pkey, nil
 }
+
+// encodeEcPoint encodes pt.
+func encodeEcPoint(group C.GO_EC_GROUP_PTR, pt C.GO_EC_POINT_PTR) ([]byte, error) {
+	// Get encoded point size.
+	n := C.go_openssl_EC_POINT_point2oct(group, pt, C.GO_POINT_CONVERSION_UNCOMPRESSED, nil, 0, nil)
+	if n == 0 {
+		return nil, newOpenSSLError("EC_POINT_point2oct")
+	}
+	// Encode point into bytes.
+	bytes := make([]byte, n)
+	n = C.go_openssl_EC_POINT_point2oct(group, pt, C.GO_POINT_CONVERSION_UNCOMPRESSED, base(bytes), n, nil)
+	if n == 0 {
+		return nil, newOpenSSLError("EC_POINT_point2oct")
+	}
+	return bytes, nil
+}
+
+// deriveAndEncode calls derive to generate a public key point and then encodes it.
+func deriveAndEncode(nid C.int, derive func(group C.GO_EC_GROUP_PTR) (C.GO_EC_POINT_PTR, error)) ([]byte, error) {
```

Comment:
The term "derive" confuses me a little bit here.

Do I read it right that this is referring to just extracting a public key from the private key?

Or is this ECDH or other kind of actual key derivation ?

Reply:
Good point. This function generates and encodes a public key, it doesn't derive anything. Will rename it to `generateAndEncodePublicKey`.